src/nvim/os/stdpaths.c

```diff
@@ -0,0 +1,100 @@
+#include "nvim/os/os.h"
+#include "nvim/strings.h"
+#include "nvim/path.h"
+#include "nvim/garray.h"
+
+typedef enum {
+  kConfigHome,
+  kDataHome,
+  kCacheHome,
+  kRuntimeDir,
+  kConfigDirs,
+  kDataDirs,
+} XDGDirType;
+
+static const char *xdg_env_vars[] = {
+  [kConfigHome] = "XDG_CONFIG_HOME",
+  [kDataHome] = "XDG_DATA_HOME",
+  [kCacheHome] = "XDG_CACHE_HOME",
+  [kRuntimeDir] = "XDG_RUNTIME_DIR",
+  [kConfigDirs] = "XDG_CONFIG_DIRS",
+  [kDataDirs] = "XDG_DATA_DIRS",
+};
+
+static const char *const xdg_defaults[] = {
+  // Windows, Apple stuff are just shims right now
+#ifdef WIN32
+  // Windows
+#elif APPLE
+  // Apple (this includes iOS, which we might need to handle differently)
+  [kConfigHome] = "~/Library/Preferences",
+  [kDataHome] = "~/Library/Application Support",
+  [kCacheHome] = "~/Library/Caches",
+  [kRuntimeDir] = "~/Library/Application Support",
+  [kConfigDirs] = "/Library/Application Support",
+  [kDataDirs] = "/Library/Application Support",
+#else
+  // Linux, BSD, CYGWIN
+  [kConfigHome] = "~/.config",
+  [kDataHome] = "~/.local/share",
+  [kCacheHome] = "~/.cache",
+  [kRuntimeDir] = "",
+  [kConfigDirs] = "/etc/xdg/",
+  [kDataDirs] = "/usr/local/share/:/usr/share/",
```

Comment:
Picking up the question about tailing slashes again. 

I see what you mean about following the spec. But I think "equal" in the spec can also mean "functional equivalent" (the spec doesn't tell you which programming language or encoding to use for example). And depending on which functions in nvim use these values it might be an advantage to have all the defaults in the same style (I would prefer no tailing slash). Otherwise the code that uses these values has to decide if it will append a slash or not. 

I am not sure if this concern will actually have any functional relevance. All applications that I now of can work with non normalized path (double slashes) when *using* files. It is only about aesthetics when the values will be displayed to the user.